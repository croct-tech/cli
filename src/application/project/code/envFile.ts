import {FileSystem} from '@/application/fs/fileSystem';

export class EnvFile {
    private readonly fileSystem: FileSystem;

    private readonly path: string;

    public constructor(fileSystem: FileSystem, path: string) {
        this.fileSystem = fileSystem;
        this.path = path;
    }

    public getName(): string {
        return this.fileSystem.getBaseName(this.path);
    }

    public exists(): Promise<boolean> {
        return this.fileSystem.exists(this.path);
    }

    public async hasVariable(name: string): Promise<boolean> {
        const content = await this.read();

        if (content === '') {
            return false;
        }

        const escapedVariable = EnvFile.escapeRegex(name);
        const regex = new RegExp(`^${escapedVariable}\\s*=`, 'm');

        return regex.test(content);
    }

    public async setVariables(variables: Record<string, string>): Promise<void> {
        const content = await this.read();

        try {
            for (const [name, value] of Object.entries(variables)) {
                await this.setVariable(name, value);
            }
        } catch (error) {
            await this.write(content);

            throw error;
        }
    }

    public async setVariable(name: string, value: string): Promise<void> {
        const content = await this.read();
        const escapedName = EnvFile.escapeRegex(name);

        if (content === '') {
            return this.write(`${name}=${value}`);
        }

        const entry = `${name}=${value}`;

        const updatedContent = content.replace(
            new RegExp(
                `${escapedName}\\s*=\\s*((?!['"\`]).*$|\`(?:\\.|[^\`])*\`|'(?:\\.|[^'])*'|"(?:\\.|[^"])*")`,
                'm',
            ),
            entry,
        );

        if (updatedContent !== content || updatedContent.includes(entry)) {
            return this.write(updatedContent);
        }

        return this.write(`${content}${content.endsWith('\n') ? '' : '\n'}${name}=${value}`);
    }

    private async write(content: string): Promise<void> {
        await this.fileSystem.writeTextFile(this.path, content, {
            overwrite: true,
        });
    }

    private async read(): Promise<string> {
        if (!await this.exists()) {
            return '';
        }

        return this.fileSystem.readTextFile(this.path);
    }

    private static escapeRegex(value: string): string {
        return value.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
    }
}
