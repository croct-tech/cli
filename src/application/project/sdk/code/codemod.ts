export type ResultCode<T> = {
    modified: boolean,
    result: T,
};

export type CodemodOptions = Record<string, any>;

export interface Codemod<I, O extends CodemodOptions = CodemodOptions> {
    apply(input: I, options?: O): Promise<ResultCode<I>>;
}

export class CodemodError extends Error {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, CodemodError.prototype);
    }
}

export class MalformedCodeError extends CodemodError {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, MalformedCodeError.prototype);
    }
}
