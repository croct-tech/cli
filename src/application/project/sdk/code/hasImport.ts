import {parse} from '@babel/parser';
import {visit} from 'recast';
import {namedTypes as Ast} from 'ast-types/gen/namedTypes';

export type ImportMatcher = {
    moduleName: string|RegExp,
    importName?: string|RegExp,
};

export function hasImport(source: string, matcher: ImportMatcher): boolean {
    let ast: Ast.Node;

    try {
        ast = parse(source, {
            sourceType: 'module',
            plugins: ['jsx', 'typescript'],
        });
    } catch {
        return false;
    }

    let found = false;

    visit(ast, {
        visitImportDeclaration: function accept(path) {
            const {node} = path;

            if (!Ast.StringLiteral.check(node.source) || !matches(node.source.value, matcher.moduleName)) {
                return false;
            }

            if (matcher.importName === undefined) {
                found = true;

                return false;
            }

            for (const specifier of node.specifiers ?? []) {
                if (
                    Ast.ImportSpecifier.check(specifier)
                    && Ast.Identifier.check(specifier.imported)
                    && matches(specifier.imported.name, matcher.importName)
                ) {
                    found = true;

                    return false;
                }
            }

            return false;
        },
    });

    return found;
}

function matches(value: string, matcher: string|RegExp): boolean {
    if (typeof matcher === 'string') {
        return value === matcher;
    }

    return matcher.test(value);
}
