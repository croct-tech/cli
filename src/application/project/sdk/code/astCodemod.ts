import type {namedTypes as Ast} from 'ast-types';
import {print} from 'recast';
import {Codemod, ResultCode} from '@/application/project/sdk/code/transformation';
import {parse} from '@/application/project/sdk/code/parser';

export class AstCodemod implements Codemod<string> {
    private readonly codemods: Array<Codemod<Ast.Node>>;

    public constructor(...transformers: Array<Codemod<Ast.Node>>) {
        this.codemods = transformers;
    }

    public apply(input: string): ResultCode<string> {
        if (this.codemods.length === 0) {
            return {
                modified: false,
                result: input,
            };
        }

        const ast = parse(input, ['jsx', 'typescript']);

        const result = this.codemods.reduce<ResultCode<Ast.Node>>(
            (next, transformer) => {
                const transformation = transformer.apply(next.result);

                return {
                    modified: transformation.modified || next.modified,
                    result: transformation.result,
                };
            },
            {
                modified: false,
                result: ast,
            },
        );

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result).code,
        };
    }
}
