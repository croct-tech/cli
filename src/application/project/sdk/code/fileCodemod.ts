import {readFile, writeFile} from 'fs/promises';
import {Codemod, CodemodError, ResultCode} from '@/application/project/sdk/code/transformation';
import {formatCause} from '@/application/error';

export class FileCodemod implements Codemod<string> {
    private readonly codemod: Codemod<string>;

    public constructor(codemod: Codemod<string>) {
        this.codemod = codemod;
    }

    public async apply(input: string): Promise<ResultCode<string>> {
        let source = '';

        try {
            source = await readFile(input, 'utf-8');
        } catch (error) {
            if (error.code !== 'ENOENT') {
                throw new CodemodError(`Failed to read file: ${formatCause(error)}`);
            }
        }

        const result = await this.codemod.apply(source);

        if (result.modified) {
            try {
                await writeFile(input, result.result, {encoding: 'utf-8', flag: 'w'});
            } catch (error) {
                throw new CodemodError(`Failed to write file: ${formatCause(error)}`);
            }
        }

        return result;
    }
}
