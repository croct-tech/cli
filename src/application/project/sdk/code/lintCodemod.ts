import {Codemod, ResultCode} from '@/application/project/sdk/code/transformation';
import {Linter} from '@/application/project/linter';

export class LintCodemod implements Codemod<string> {
    private readonly codemod: Codemod<string>;

    private readonly linter: Linter;

    public constructor(codemod: Codemod<string>, linter: Linter) {
        this.codemod = codemod;
        this.linter = linter;
    }

    public async apply(input: string): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(input);

        if (result.modified) {
            await this.linter.fix([input]);
        }

        return result;
    }
}
