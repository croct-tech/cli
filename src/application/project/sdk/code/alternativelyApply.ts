import {Codemod, CodemodOptions, ResultCode} from '@/application/project/sdk/code/codemod';

export class AlternativelyApply<T, O extends CodemodOptions> implements Codemod<T, O> {
    private readonly codemods: Array<Codemod<T, O>>;

    public constructor(codemod: Codemod<T, O>, ...codemods: Array<Codemod<T, O>>) {
        this.codemods = [codemod, ...codemods];
    }

    public async apply(input: T, options?: O): Promise<ResultCode<T>> {
        for (const codemod of this.codemods) {
            const result = await codemod.apply(input, options);

            if (result.modified) {
                return result;
            }
        }

        return {
            modified: false,
            result: input,
        };
    }
}
