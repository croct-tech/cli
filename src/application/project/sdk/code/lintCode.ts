import {Codemod, CodemodOptions, ResultCode} from '@/application/project/sdk/code/codemod';
import {Linter} from '@/application/project/linter';

export class LintCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<string, O>;

    private readonly linter: Linter;

    public constructor(codemod: Codemod<string, O>, linter: Linter) {
        this.codemod = codemod;
        this.linter = linter;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(input, options);

        if (result.modified) {
            await this.linter.fix([input]);
        }

        return result;
    }
}
