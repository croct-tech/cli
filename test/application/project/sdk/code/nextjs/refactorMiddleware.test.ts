import {resolve} from 'path';
import {MiddlewareOptions, RefactorMiddleware} from '@/application/project/sdk/code/nextjs/refactorMiddleware';
import {loadFixtures} from '../fixtures';
import {AstCodemod} from '@/application/project/sdk/code/astCodemod';

describe('RefactorMiddleware', () => {
    const defaultOptions: MiddlewareOptions = {
        import: {
            module: '@croct/plug-next/middleware',
            functionName: 'withCroct',
            matcherName: 'matcher',
            matcherLocalName: 'matcher',
        },
    };

    const scenarios = loadFixtures<MiddlewareOptions>(
        resolve(__dirname, '../fixtures/middleware'),
        defaultOptions,
        {
            'matcherAlias.ts': {
                import: {
                    ...defaultOptions.import,
                    matcherLocalName: 'croctMatcher',
                },
            },
        },
    );

    it.each(scenarios)('should correctly transform $name', ({name, fixture, options}) => {
        const transformer = new AstCodemod(new RefactorMiddleware(options));
        const output = transformer.apply(fixture);

        expect(output.result).toMatchSnapshot(name);
    });
});
