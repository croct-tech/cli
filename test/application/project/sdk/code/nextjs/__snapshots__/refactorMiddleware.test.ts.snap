// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefactorMiddleware should correctly transform configAfterDefaultExport.ts: configAfterDefaultExport.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request: NextRequest) {
        console.log(request.url);
    }
});
"
`;

exports[`RefactorMiddleware should correctly transform configAfterMiddlewareWithReference.ts: configAfterMiddlewareWithReference.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

// pattern
const PATTERN = '/((?!api|_next/static|_next/image|favicon.ico).*)';

// matcher
class Matcher {
    static readonly PATTERN = PATTERN;
}

// getMatcher
function getMatcher() {
    return Matcher.PATTERN;
}

// currentMatcher
const currentMatcher = getMatcher();

// currentConfig
const currentConfig = {
    matcher: [
        ...(Array.isArray(currentMatcher) ? currentMatcher : [currentMatcher]),
        matcher
    ],
}

// middleware
export const middleware = withCroct({
    matcher: currentConfig.matcher,

    next: function(request: NextRequest) {
        console.log(request.url);
    }
});

function unrelated() {
}

// Non-global scope
{
    // Different scope, should not be considered
    const PATTERN = null;

    function getMatcher() {
    }

    class Matcher {
    }
}

// config
export const config = currentConfig;
"
`;

exports[`RefactorMiddleware should correctly transform configAfterNamedExport.ts: configAfterNamedExport.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export default withCroct({
    matcher: config.matcher,

    next: function middleware(request: NextRequest) {
        console.log(request.url);
    }
});
"
`;

exports[`RefactorMiddleware should correctly transform configInvalidReference.ts: configInvalidReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export default withCroct(function middleware(request: NextRequest) {
    console.log(request.url);
});

export const config = bar;
"
`;

exports[`RefactorMiddleware should correctly transform configWithArrayMatcher.ts: configWithArrayMatcher.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request: NextRequest) {
        console.log(request.url);
    }
});
"
`;

exports[`RefactorMiddleware should correctly transform configWithIndirectVariableReference.ts: configWithIndirectVariableReference.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
const regex = '/((?!api|_next/static|_next/image|favicon.ico).*)';

const configValue = {
    matcher: [...(Array.isArray(regex) ? regex : [regex]), matcher],
}

const indirectReference = configValue;

export const config = indirectReference;

export const middleware = withCroct({
    matcher: configValue.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`RefactorMiddleware should correctly transform configWithStringMatcher.ts: configWithStringMatcher.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`RefactorMiddleware should correctly transform configWithVariableMatcher.ts: configWithVariableMatcher.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
const regex = '/((?!api|_next/static|_next/image|favicon.ico).*)';

export const config = {
    matcher: [...(Array.isArray(regex) ? regex : [regex]), matcher],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`RefactorMiddleware should correctly transform configWithVariableReference.ts: configWithVariableReference.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
const regex = '/((?!api|_next/static|_next/image|favicon.ico).*)';

const configValue = {
    matcher: [...(Array.isArray(regex) ? regex : [regex]), matcher],
}

export const config = configValue;

export const middleware = withCroct({
    matcher: configValue.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`RefactorMiddleware should correctly transform configWithoutMatcher.ts: configWithoutMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export const config = {
}

export const middleware = withCroct(function(request) {
    console.log(request.url);
});
"
`;

exports[`RefactorMiddleware should correctly transform defaultExportAnonymousFunction.ts: defaultExportAnonymousFunction.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(function() {
    console.log('middleware');
});
"
`;

exports[`RefactorMiddleware should correctly transform defaultExportArrowFunction.ts: defaultExportArrowFunction.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(() => {
    console.log('middleware');
});
"
`;

exports[`RefactorMiddleware should correctly transform defaultExportArrowFunctionReference.ts: defaultExportArrowFunctionReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const anything = withCroct((request) => console.log(request.url));

export default anything;
"
`;

exports[`RefactorMiddleware should correctly transform defaultExportArrowFunctionWithBodyReference.ts: defaultExportArrowFunctionWithBodyReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const anything = withCroct((request) => {
    console.log(request.url);
})

export default anything;
"
`;

exports[`RefactorMiddleware should correctly transform defaultExportClass.ts: defaultExportClass.ts 1`] = `
"export default class Middleware {
    // invalid
}
"
`;

exports[`RefactorMiddleware should correctly transform defaultExportFunctionDeclaration.ts: defaultExportFunctionDeclaration.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(function anything(request) {
    console.log(request.url);
});
"
`;

exports[`RefactorMiddleware should correctly transform defaultExportFunctionDeclarationReference.ts: defaultExportFunctionDeclarationReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

const anything = withCroct(function(request) {
    console.log(request.url);
});

export default anything;
"
`;

exports[`RefactorMiddleware should correctly transform defaultExportFunctionExpressionReference.ts: defaultExportFunctionExpressionReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const anything = withCroct(function (request) {
    console.log(request.url);
})

export default anything;
"
`;

exports[`RefactorMiddleware should correctly transform defaultExportIndirectReference.ts: defaultExportIndirectReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const anything = function (request) {
    console.log(request.url);
}

const something = withCroct(anything);

export default something;
"
`;

exports[`RefactorMiddleware should correctly transform empty.ts: empty.ts 1`] = `""`;

exports[`RefactorMiddleware should correctly transform matcherAlias.ts: matcherAlias.ts 1`] = `
"import { withCroct, matcher as croctMatcher } from "@croct/plug-next/middleware";

export const config = {
    matcher: ['.*', croctMatcher]
};

export const middleware = withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
});
"
`;

exports[`RefactorMiddleware should correctly transform namedExportArrowFunction.ts: namedExportArrowFunction.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export const middleware = withCroct(() => console.log('middleware'));
"
`;

exports[`RefactorMiddleware should correctly transform namedExportArrowFunctionWithBody.ts: namedExportArrowFunctionWithBody.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export const middleware = withCroct(() => {
    console.log('middleware');
})
"
`;

exports[`RefactorMiddleware should correctly transform namedExportFunctionDeclaration.ts: namedExportFunctionDeclaration.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const middleware = withCroct(function() {
    console.log('middleware');
});
"
`;

exports[`RefactorMiddleware should correctly transform namedExportFunctionExpression.ts: namedExportFunctionExpression.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export const middleware = withCroct(function() {
    console.log('middleware');
})
"
`;

exports[`RefactorMiddleware should correctly transform namedSpecifiedExport.ts: namedSpecifiedExport.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";

const config = {
    matcher: ['.*', matcher]
};

const middleware = withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
})

export { middleware, config };
"
`;

exports[`RefactorMiddleware should correctly transform specifiedExportWithAliases.ts: specifiedExportWithAliases.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
function unrelated() {
}

const _config = {
    matcher: ['.*', matcher]
};

const _middlewareFn = withCroct({
    matcher: _config.matcher,

    next: function() {
        console.log('middleware');
    }
})

export {
    _middlewareFn as default,
    _config as config
};
"
`;

exports[`RefactorMiddleware should correctly transform unrelatedExports.ts: unrelatedExports.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export function foo() {
}

export default withCroct(() => {

});

export function middleware() {
}
"
`;
